src/traefik.py

```diff
@@ -0,0 +1,616 @@
+
+        else:
+            transports = {}
+
+        config = {
+            "http": {
+                "routers": router_cfg,
+                "services": {traefik_service_name: service_def},
+            },
+        }
+        # Traefik does not accept an empty serversTransports. Add it only if it's non-empty.
+        if transports:
+            config["http"].update({"serversTransports": transports})
+
+        middlewares = self._generate_middleware_config(
+            redirect_https=redirect_https_,
+            strip_prefix=strip_prefix_,
+            scheme=scheme_,
+            prefix=prefix,
+        )
+
+        if middlewares:
+            config["http"]["middlewares"] = middlewares
+            router_cfg[traefik_router_name]["middlewares"] = list(middlewares.keys())
+
+            if f"{traefik_router_name}-tls" in router_cfg:
+                router_cfg[f"{traefik_router_name}-tls"]["middlewares"] = list(middlewares.keys())
+
+        return config
+
+    def _generate_middleware_config(
+        self,
+        redirect_https: bool,
+        strip_prefix: bool,
+        scheme: str,
+        prefix: str,
+    ) -> dict:
+        """Generate a middleware config.
+
+        We need to generate a different section per middleware type, otherwise traefik complains:
+          "cannot create middleware: multi-types middleware not supported, consider declaring two
+          different pieces of middleware instead"
+        """
+        no_prefix_middleware = {}  # type: Dict[str, Dict[str, Any]]
+        if self._routing_mode is RoutingMode.path:
+            if strip_prefix:
+                no_prefix_middleware[f"juju-sidecar-noprefix-{prefix}"] = {
+                    "stripPrefix": {"prefixes": [f"/{prefix}"], "forceSlash": False}
+                }
+
+        # Condition rendering the https-redirect middleware on the scheme, otherwise we'd get a 404
+        # when attempting to reach an http endpoint.
+        redir_scheme_middleware = {}
+        if redirect_https and scheme == "https":
+            redir_scheme_middleware[f"juju-sidecar-redir-https-{prefix}"] = {
+                "redirectScheme": {"scheme": "https", "port": 443, "permanent": True}
+            }
+
+        return {**no_prefix_middleware, **redir_scheme_middleware}
+
+    @staticmethod
+    def generate_tls_config_for_route(
+        router_name: str,
+        route_rule: str,
+        service_name: str,
+        external_host: str,
+    ) -> Dict[str, Any]:
+        """Generate a TLS configuration segment."""
+        if is_hostname(external_host):
+            tls_entry = {
+                "domains": [
+                    {
+                        "main": external_host,
+                        "sans": [f"*.{external_host}"],
+                    },
+                ],
+            }
+
+        else:
+            # When the external host is a bare IP, we do not need the 'domains' entry.
+            tls_entry = {}
```

Comment:
The `else` clause can be omitted.

```suggestion
        tls_entry = {}

        if is_hostname(external_host):
            # When the external host is not a bare IP, we need the 'domains' entry.
            tls_entry = {
                "domains": [
                    {
                        "main": external_host,
                        "sans": [f"*.{external_host}"],
                    },
                ],
            }
```

Reply:
in this case, I think "explicit is better than implicit" takes precedence. tls is complicated already :)